pkg/api/status/v1.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"fmt"
+	"math"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/route"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/tsdb"
+	"github.com/prometheus/prometheus/tsdb/index"
+	v1 "github.com/prometheus/prometheus/web/api/v1"
+	"github.com/thanos-io/thanos/pkg/api"
+	extpromhttp "github.com/thanos-io/thanos/pkg/extprom/http"
+	"github.com/thanos-io/thanos/pkg/logging"
+)
+
+// Stat holds the information about individual cardinality.
+type Stat struct {
+	Name  string `json:"name"`
+	Value uint64 `json:"value"`
+}
+
+func convertStats(stats []index.Stat) []Stat {
+	result := make([]Stat, 0, len(stats))
+	for _, item := range stats {
+		item := Stat{Name: item.Name, Value: item.Count}
+		result = append(result, item)
+	}
+	return result
+}
+
+// TSDBStatus has information of cardinality statistics from postings.
+type TSDBStatus struct {
+	HeadStats                   v1.HeadStats `json:"headStats"`
+	SeriesCountByMetricName     []Stat       `json:"seriesCountByMetricName"`
+	LabelValueCountByLabelName  []Stat       `json:"labelValueCountByLabelName"`
+	MemoryInBytesByLabelName    []Stat       `json:"memoryInBytesByLabelName"`
+	SeriesCountByLabelValuePair []Stat       `json:"seriesCountByLabelValuePair"`
+}
+
+type GetStatsFunc func(r *http.Request, statsByLabelName string) (*tsdb.Stats, error)
+
+type Options struct {
+	GetStats GetStatsFunc
+	Registry *prometheus.Registry
+}
+
+type StatusAPI struct {
+	getTSDBStats GetStatsFunc
+	options      Options
+}
+
+func New(opts Options) *StatusAPI {
+	return &StatusAPI{
+		getTSDBStats: opts.GetStats,
+		options:      opts,
+	}
+}
+
+func (sapi *StatusAPI) Register(r *route.Router, tracer opentracing.Tracer, logger log.Logger, ins extpromhttp.InstrumentationMiddleware, logMiddleware *logging.HTTPServerMiddleware) {
+	instr := api.GetInstr(tracer, logger, ins, logMiddleware, false)
+	r.Get("/api/v1/status/tsdb", instr("tsdb_status", sapi.httpServeStats))
```

Comment:
Other federated APIs like rules, targets, exemplars are exposed as gRPC APIs at the store, and federations are done at the Querier level.
For this API do we have any plan to implement the gRPC version of it and then aggregated at the Query level?

Reply:
So we have this issue at the moment: https://github.com/thanos-io/thanos/issues/5395

Maybe we can discuss federation there, but I think it can be a good idea to do it. We just have to see how merging is going to work, and whether we want to merge the results in a single json object in the result, or have one object per store target. I think this is a question for getting tsdb stats for all tenants as well.

Reply:
Regarding merging, I'm wondering how data replication for Receive will affect it. AFAIU, we use quorum-based logic for replication, so maybe merging this would lead to some inaccuracies in TSDB statistics for a tenant?

Reply:
Yes agreed. The stats can get very inaccurate if we merge across replicas.